### PR TITLE
When setting rule IDs, calculate which tokens are missing from lexer / parser.

### PR DESCRIPTION
**Pull request by ltratt**
Tokens defined in the lexer but not referenced in the parser are quite probably
harmless (or, at most, worth warning the user about). But tokens referenced in the parser but not defined in the lexer are highly indicative of an error. Calculating these and returning them to the caller allows them to make a choice as to whether they want to do something with this information or not.

A corresponding PR for lrpar (based on https://github.com/softdevteam/lrpar/commit/1af1a8b1b76d907c89bad54ebd0649fdf74510ff) will make use of this functionality. [I don't know if rstreediff needs to do anything with this information -- that's up to you.]

### CODE REVIEW

File: src/lib/lexer.rs [hunk 1]
```diff
@@ -81,16 +81,73 @@ impl<TokId: Copy + Eq> LexerDef<TokId> {
...
     /// Set the id attribute on rules to the corresponding value in `map`. This is typically used
-    /// to synchronise a parser's notion of token IDs with the lexers.
-    pub fn set_rule_ids(&mut self, map: &HashMap<&str, TokId>) {
-        for r in self.rules.iter_mut() {
+    /// to synchronise a parser's notion of token IDs with the lexers. While doing this, it keeps
+    /// track of which tokens:
+    ///   1) are defined in the lexer but not referenced by the parser
+    ///   2) and referenced by the parser but not defined in the lexer
+    /// and returns them as a tuple `(Option<HashSet<&str>>, Option<HashSet<&str>>)` in the order
+    /// (defined_in_lexer_missing_from_parser, referenced_in_parser_missing_from_lexer). Since in
```
Comment (snim2) [inline src/lib/lexer.rs]: Backticks missing around names?

Comment (ltratt) [reply src/lib/lexer.rs]: They're not variables, so I'm not sure they should be backticked (I just couldn't work out how to find labels that wouldn't be confusing, so this was my best attempt -- suggestions welcome!).

Comment (snim2) [reply src/lib/lexer.rs]: Ah, I see. I was expecting to see those names in the block below! Maybe make them italic?

```markdown
(*defined-in-lexer-missing-from-parser*, *referenced-in-parser-missing-from-lexer*)
```

Comment (ltratt) [reply src/lib/lexer.rs]: Ok, try the new commit.

**Comment by snim2** [comment]
Looks sensible. If you are happy then I think this is ready to squash.

**Comment by ltratt** [comment]
OK, squashed.

**Pull request merged**